### PR TITLE
Package management update and documentation / testing

### PR DESCRIPTION
**Pull request by slobentanzer**
Switched to Poetry for package management, depending on the way the package is used in pipelines may need some testing. Incremented minor version to 2 due to this change.

Added docstrings for all functions in the _pipeline module and tests for most. Some tests are not functional yet.

### CODE REVIEW

**Comment by ktpolanski** [comment]
If I understand correctly, the QC is only a part of this repository here, and your plan is to update it so it's properly documented, installable from pip, and just generally more useable. Correct?

Do you have plans for non-QC functions in this repository? Is it okay to leave them as is as you work on the parts relevant to you?

**Comment by slobentanzer** [comment]
Hi @ktpolanski,
yes, it is OK to leave them as is, and ultimately, IMO, the entire package should be properly documented, including all the other modules. It is not my decision what should be part of the sctk package, but I would suggest to consistently document all parts to make it easier for the prospective users to find their way around it.

**Comment by ktpolanski** [comment]
How would you feel about spinning off the QC stuff into its own repository for further development? Would having it start as a fork of this one here grant nh3 (and whoever else) sufficient credit?

**Comment by slobentanzer** [comment]
After talking to Amanda, the plan for my involvement was to streamline this repository for dissemination and usage. Maybe you should clarify that internally; my main aim here was to make the package more usable.

If I wanted to simply fork the QC part, I could already do so, according to open-source etiquette. It is not my main aim with this PR. Generally, I believe that maintaining one central version of a program, with contributions from others in the ecosystem, is the right way in open source / open science projects.

**Comment by ktpolanski** [comment]
Sure thing, understandable. Thanks for this!

I took a look at the PR and it looks good - there are docstrings and tests and whatnot. How would one actually run the tests in the current form? Can this be automated GitHub side?

Sorry about the questions, I'm not particularly GitHub savvy.

**Comment by slobentanzer** [comment]
At the moment, you can run them using pytest in the cloned repository on your machine, but yes, I am planning to make this a CI (continuous integration) step using a GitHub action. Then, the tests would automatically run on every push to the repository.

**Comment by ktpolanski** [comment]
Great, thanks. I created a fresh conda environment with python 3.9, cloned your fork, `pip install .`'ed it, and then also pipped in pytest and tried to run the tests. Seems two failed, and then I also see some mentions of nans in other areas?

https://pastebin.com/uFxkazQn

**Comment by slobentanzer** [comment]
Yes, that is correct. The tests failing is not a reflection on the pipeline, but on the incompleteness of the tests. Amanda has put me in contact with Ni and Batu, and with their knowledge we should be able to fix the tests to be more relevant to the pipeline functions. The aim is to facilitate further developments by encapsulating the current functionality in the tests, to make sure new additions don't break existing functionality.

**Comment by NickEngland** [comment]
Hi @slobentanzer ,

You've got some great changes here for the documentation and tests, would you like to develop these against the dev branch on our github? We can change the pull request target to dev and merge.

I'm a little hesitant to have tests on the master branch which don't pass, as that can be confusing for people who are just trying to check out the package. Or we can keep this pull request open and merge it after you've finished? Which option would you prefer?

**Comment by slobentanzer** [comment]
Hi @NickEngland,
absolutely agree that we should not merge non-passing tests. I'd keep the PR open for now, because I don't know what exactly the `dev` branch is used for. I tend to favour creation of single short-lived feature branches and am not a fan of long-running development branches, as those often add confusion and maintenance effort.

As an aside, the `dev` branch seems to be inactive for almost a year and 5 commits behind main (with no commits ahead), so if it were my repo, I would probably retire that branch. Similar for `deviance`.

**Comment by ktpolanski** [comment]
`deviance` was me contributing a function and is now deleted

**Comment by slobentanzer** [comment]
Hi @NickEngland,
all tests pass now. Even though I am fairly sure they do not yet cover all functional aspects of the QC, at least they should not confuse new users checking out the package.

I'd like to continue with the CI/CD part based on these initial tests, so it would be great from my end if we could merge this PR. With triage rights I currently cannot create branches on this repo, and multiple open PRs could become confusing. Now that I think of it, I am also not entirely sure if I can manage GitHub Actions with triage permissions. I'll get back to you on that.

**Comment by slobentanzer** [comment]
Of note, before merging: I fixed what I perceived to be a bug in commit 2835d2c25dd2088365fc0f7d168fdc6e1431a810, but that could also result from the test being implemented differently from how the pipeline is normally used, please check this before merging into `main`.

File: sctk/_pipeline.py
Comment (slobentanzer) [reply sctk/_pipeline.py:86]: yes, you're right. the examples are just boilerplate as of now, they also need to be checked if they make sense, content-wise (same as the tests).

Comment (slobentanzer) [reply sctk/_pipeline.py]: fixed in 496393e5264f0ba88ce17ed32305131651509f2b

**Comment by ktpolanski** [comment]
Nick bumped the version and bundled the existing code into a v0.1.1 release on GitHub for archival purposes. This created a merge conflict, as now `setup.py` was one character different.

I had little luck with official GitHub command line conflict resolution instructions, as for whatever reason you seemed to be absent from the commit log. I remembered a collaborator being able to write to my fork when I submitted a PR, and found [these instructions](https://stackoverflow.com/questions/38949951/how-to-solve-merge-conflicts-across-forks) for handling the conflict. I ran `git rm setup.py` once it spawned that file back in as I tried to merge, then added the strange empty commit and pushed.

**Comment by slobentanzer** [comment]
Good job! I saw that it was weird; since I deleted `setup.py`, I also couldn't make sense of the conflict. Good thing it seems to be solved by your empty commit.

Who will be maintaining the PyPI package? Do you have an account there already? This is ideally someone who can commit to longer-term maintenance of the package. It is not technically difficult; once the account is set up, the package can be released with `poetry build` followed by `poetry publish` (can also be automated for new releases).

File: test/test_pipeline.py
Comment (ktpolanski) [inline test/test_pipeline.py:138]: the `key_added` is used when constructing the output `.obs`, and this is actually under `"custom_good_qc_clusters_fqo2"`

Comment (ktpolanski) [reply test/test_pipeline.py:138]: I rejiggered the code a bit on my end and this will need a different change now

Comment (slobentanzer) [reply test/test_pipeline.py:138]: do you want to commit your changes to the PR, or is this just a note to yourself?

Comment (ktpolanski) [reply test/test_pipeline.py:138]: I'm putting finishing touches on a demo notebook and will push some stuff relatively soon. Sorry about this, should have been clearer, but messing around with things in practice gave me a couple ideas and I let things flow.

**Comment by ktpolanski** [comment]
Master list of altered stuff:
 - Demo notebook, left in the root directory for now. I'd imagine it will get relocated somewhere once docs take shape.
 - `filter_qc_outlier2()` is now `filter_qc_outlier()` and explicitly part of the workflow. The computed calls get saved to the input object rather than output as a numpy array.
 - The emergency backup simple filtering in case `filter_qc_outlier()` fails relocated to that function from `generate_qc_clusters()`
 - Slight argument rework for the affected functions, including a `**kwargs` to get passed to the GMM.
 - Slight docstring amendments.

Affected tests are almost certainly busted, but the general idea here is to propose a user-minded analysis flow. And I think this accomplishes that.

I have also located Ni's original `qc.py` script on the farm, so I have an actual concrete example of a more advanced application of the pipeline to desouped and spliced data if desired.

**Comment by slobentanzer** [comment]
@ktpolanski very good; I agree that the workflow should be first priority. The tests need to be adjusted accordingly. I propose to make that a tightly related part of development. I have, on a different branch, already set up a basic CI workflow that we can run on each new PR to make sure existing functionality is consistent. Would suggest to merge that once we are happy with the first version of this PR.

Have a good weekend!

**Comment by ktpolanski** [comment]
This is actually an algorithmic enhancement - while trying to comprehend the innards of the QC workflow to write it up in the demo notebook, I realised that the GMM part could use automatic component selection. As such, this has now been implemented. Running with `n_components=[10]` will result in behaviour analogous to what prior code did.

At this point I think we need to clean up the docstrings for the pertinent functions and ensure the tests are up to your desired standards.

**Comment by ktpolanski** [comment]
I updated the pertinent docstrings to reflect my understanding of the code, and made sure the tests are up to date. I removed the `force` test, and there's an argument to be made for removing the `force` parameter altogether as a similar behaviour can be accomplished by controlling the last element of the metric parameter tuple.

I barely engaged with functions outside of the main QC workflow. I updated a `fit_gaussian()` call to reflect the move to `n_components`, and one of the workflow functions made use of the original outlier filter function. These were not used in the QC workflow that was ran in Amanda's paper, and the current demo notebook is a good demonstration of how to use that functionality.

I am happy with the current state of things.

**Review by ktpolanski — APPROVED** [review]

**Comment by slobentanzer** [comment]
@ktpolanski sorry I may have automatically dismissed your review by adding more commits. I just checked your code and all tests pass. The notebook also works, great explanation. I moved it to the notebooks directory and gave it a more informative name, in anticipation of other notebooks to follow.

Could you approve again?

**Review by ktpolanski — DISMISSED** [review]

**Comment by ktpolanski** [comment]
I believe this addresses your various issues.

**Comment by slobentanzer** [comment]
@ktpolanski yes; I was actually planning for it to be more of a discussion, but that's also good. ;)

The notebook still may need to be updated to reflect the renaming.

**Comment by ktpolanski** [comment]
Why is this commit not showing up in this PR? https://github.com/slobentanzer/sctk/commit/25304103d1f8263c40682d60d075e1c297b36012

**Comment by slobentanzer** [comment]
I am not sure, I followed the link in the email and GitHub told me it cannot find the commit. I was able to pull the code into my local repo just fine.

**Comment by slobentanzer** [comment]
Seems it also survived the merge as well. GitHub quirks I guess.

**Comment by ktpolanski** [comment]
I merged, I went to look at the notebook, it was not reflecting the changes. I recreated the commit and now everything is up to spec.

**Pull request merged**
